src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -271,7 +307,7 @@ public IList Get(QueryKey key, ICacheAssembler[] returnTypes, bool isNaturalKeyL
 
 				for (var i = 0; i < keys.Length; i++)
 				{
-					if (finalReturnTypes[i] == null)
+					if (cacheables[i] == null)
```

Comment:
Checking for null return types as a marker of cache miss is wrong for empty result lists with type auto-discovery: their return types are null, but that is no cache miss. So instead, we need to nullify the cacheable when it is a cache miss due to its timestamp.